File: crates/catalog/src/queries/variant_table.rs
use crate::errors::{CatalogError, CatalogResult};
use serde_json::{json, Value};
use sqlx::{PgPool, Row};
use std::collections::HashMap;

#[derive(Debug, Clone, Default)]
pub struct VariantTableQuery {
    pub filters: HashMap<String, FilterValue>,
    pub sort: Option<String>,
    pub page_size: Option<usize>,
}

#[derive(Debug, Clone)]
pub enum FilterValue {
    Equals(String),
    Range { gte: Option<f64>, lte: Option<f64> },
}

pub async fn query_variant_table(
    pool: &PgPool,
    tenant_id: &str,
    family_key: &str,
    query: &VariantTableQuery,
) -> CatalogResult<Option<Value>> {
    let row = sqlx::query(
        r#"
        SELECT payload
        FROM catalog_variant_matrix_projection
        WHERE tenant_id = $1 AND family_key = $2
        "#,
    )
    .bind(tenant_id)
    .bind(family_key)
    .fetch_optional(pool)
    .await
    .map_err(|e| CatalogError::StorageFailed(e.to_string()))?;

    let mut payload = match row {
        Some(r) => r.get::<Value, _>("payload"),
        None => return Ok(None),
    };

    let rows = payload
        .get("rows")
        .and_then(|v| v.as_array())
        .cloned()
        .unwrap_or_default();

    let mut filtered: Vec<Value> = rows
        .into_iter()
        .filter(|r| variant_matches(r, &query.filters))
        .collect();

    if let Some(sort_spec) = &query.sort {
        let (attr, dir) = parse_sort_spec(sort_spec);
        filtered.sort_by(|a, b| compare_variant(a, b, &attr, dir));
    }

    if let Some(limit) = query.page_size {
        filtered.truncate(limit);
    }

    if let Some(obj) = payload.as_object_mut() {
        obj.insert("rows".to_string(), Value::Array(filtered.clone()));
        obj.insert("rowCount".to_string(), json!(filtered.len()));
    }

    Ok(Some(payload))
}

fn variant_matches(row: &Value, filters: &HashMap<String, FilterValue>) -> bool {
    if filters.is_empty() {
        return true;
    }
    let values = match row.get("values") {
        Some(v) => v,
        None => return false,
    };
    for (attr, fv) in filters {
        let entry = match values.get(attr) {
            Some(e) => e,
            None => return false,
        };
        let raw = entry.get("raw").unwrap_or(&Value::Null);
        match fv {
            FilterValue::Equals(target) => {
                let raw_str = raw.as_str().map(|s| s.to_string()).unwrap_or_else(|| raw.to_string());
                if raw_str != *target {
                    return false;
                }
            }
            FilterValue::Range { gte, lte } => {
                let n = raw.as_f64().unwrap_or(f64::NAN);
                if let Some(g) = gte {
                    if !(n >= *g) {
                        return false;
                    }
                }
                if let Some(l) = lte {
                    if !(n <= *l) {
                        return false;
                    }
                }
            }
        }
    }
    true
}

fn parse_sort_spec(s: &str) -> (String, SortDir) {
    if let Some((attr, dir)) = s.rsplit_once('.') {
        let dir = match dir {
            "desc" => SortDir::Desc,
            _ => SortDir::Asc,
        };
        (attr.to_string(), dir)
    } else {
        (s.to_string(), SortDir::Asc)
    }
}

#[derive(Copy, Clone)]
enum SortDir {
    Asc,
    Desc,
}

fn compare_variant(a: &Value, b: &Value, attr: &str, dir: SortDir) -> std::cmp::Ordering {
    let av = a.get("values").and_then(|v| v.get(attr)).and_then(|v| v.get("normalized"));
    let bv = b.get("values").and_then(|v| v.get(attr)).and_then(|v| v.get("normalized"));

    let ord = match (av.and_then(|v| v.as_f64()), bv.and_then(|v| v.as_f64())) {
        (Some(an), Some(bn)) => an.partial_cmp(&bn).unwrap_or(std::cmp::Ordering::Equal),
        _ => {
            let as_ = av.and_then(|v| v.as_str()).unwrap_or("");
            let bs_ = bv.and_then(|v| v.as_str()).unwrap_or("");
            as_.cmp(bs_)
        }
    };
    match dir {
        SortDir::Asc => ord,
        SortDir::Desc => ord.reverse(),
    }
}

pub fn parse_filter_query(raw: &HashMap<String, String>) -> HashMap<String, FilterValue> {
    let mut out = HashMap::new();
    let mut ranges: HashMap<String, (Option<f64>, Option<f64>)> = HashMap::new();
    for (k, v) in raw {
        if let Some(rest) = k.strip_prefix("filters[") {
            if let Some(rest) = rest.strip_suffix(']') {
                if let Some((attr, op)) = rest.split_once("][") {
                    let n = v.parse::<f64>().ok();
                    let entry = ranges.entry(attr.to_string()).or_insert((None, None));
                    if op == "gte" {
                        entry.0 = n;
                    } else if op == "lte" {
                        entry.1 = n;
                    }
                } else {
                    out.insert(rest.to_string(), FilterValue::Equals(v.clone()));
                }
            }
        }
    }
    for (attr, (gte, lte)) in ranges {
        out.insert(attr, FilterValue::Range { gte, lte });
    }
    out
}

File: packages/modules-catalog/src/queries/variant-table.ts
import type { ProjectionStore } from '@atlas/ports';
import type { VariantTableParams, VariantTableResponse, FilterValue } from '@atlas/platform-core';
import { projectionKey } from '../projections/variant-matrix.ts';

interface VariantRow {
  variantId: string;
  variantKey: string;
  name: string;
  revision: number;
  values: Record<string, { raw: unknown; normalized: unknown; display: string | null }>;
}

function variantMatches(row: VariantRow, filters: Record<string, FilterValue>): boolean {
  for (const [attr, fv] of Object.entries(filters)) {
    const entry = row.values[attr];
    if (!entry) return false;
    const raw = entry.raw;
    if (fv.kind === 'equals') {
      const rawStr = typeof raw === 'string' ? raw : JSON.stringify(raw);
      if (rawStr !== fv.value) return false;
    } else {
      const n = typeof raw === 'number' ? raw : Number.NaN;
      if (fv.gte !== undefined && !(n >= fv.gte)) return false;
      if (fv.lte !== undefined && !(n <= fv.lte)) return false;
    }
  }
  return true;
}

function parseSortSpec(s: string): { attr: string; dir: 'asc' | 'desc' } {
  const idx = s.lastIndexOf('.');
  if (idx === -1) return { attr: s, dir: 'asc' };
  const attr = s.slice(0, idx);
  const dirStr = s.slice(idx + 1);
  return { attr, dir: dirStr === 'desc' ? 'desc' : 'asc' };
}

function compareVariants(a: VariantRow, b: VariantRow, attr: string): number {
  const av = a.values[attr]?.normalized;
  const bv = b.values[attr]?.normalized;
  if (typeof av === 'number' && typeof bv === 'number') return av - bv;
  const as_ = typeof av === 'string' ? av : '';
  const bs_ = typeof bv === 'string' ? bv : '';
  return as_.localeCompare(bs_);
}

export async function queryVariantTable(
  tenantId: string,
  familyKey: string,
  params: VariantTableParams,
  projections: ProjectionStore,
): Promise<VariantTableResponse | null> {
  const stored = (await projections.get(projectionKey(familyKey, tenantId))) as
    | VariantTableResponse
    | null;
  if (!stored) return null;

  const allRows = (stored.rows as unknown as VariantRow[]) ?? [];
  const filters = params.filters ?? {};
  let filtered = allRows.filter((r) => variantMatches(r, filters));

  if (params.sort) {
    const { attr, dir } = parseSortSpec(params.sort);
    filtered = [...filtered].sort((a, b) => {
      const ord = compareVariants(a, b, attr);
      return dir === 'asc' ? ord : -ord;
    });
  }

  if (params.pageSize !== undefined) {
    filtered = filtered.slice(0, params.pageSize);
  }

  return {
    ...stored,
    rows: filtered as unknown as Array<Record<string, unknown>>,
    rowCount: filtered.length,
  };
}

export function parseFilterQuery(raw: Record<string, string>): Record<string, FilterValue> {
  const out: Record<string, FilterValue> = {};
  const ranges: Record<string, { gte?: number; lte?: number }> = {};
  for (const [k, v] of Object.entries(raw)) {
    const stripped = k.startsWith('filters[') && k.endsWith(']') ? k.slice(8, -1) : null;
    if (stripped === null) continue;
    const opIdx = stripped.indexOf('][');
    if (opIdx >= 0) {
      const attr = stripped.slice(0, opIdx);
      const op = stripped.slice(opIdx + 2);
      const n = Number(v);
      if (!Number.isFinite(n)) continue;
      const entry = ranges[attr] ?? {};
      if (op === 'gte') entry.gte = n;
      else if (op === 'lte') entry.lte = n;
      ranges[attr] = entry;
    } else {
      out[stripped] = { kind: 'equals', value: v };
    }
  }
  for (const [attr, r] of Object.entries(ranges)) {
    const fv: FilterValue = { kind: 'range' };
    if (r.gte !== undefined) fv.gte = r.gte;
    if (r.lte !== undefined) fv.lte = r.lte;
    out[attr] = fv;
  }
  return out;
}
